File: nestjs/src/common/utils/writeLog.ts
import { Logger } from '@nestjs/common';
import * as fs from 'fs';
import * as path from 'path';
import yamlConfig from '../config/yaml-config.service';
import { BusinessException } from '../exceptions/business.exception';

export class LoggerUtil {
  private static logPath: string;
  private static configLoaded: boolean = false;

  // 静态初始化配置
  private static loadConfig(): void {
    if (this.configLoaded) return;

    try {
      // 复用现有的 YAML 配置服务
      const config: Record<string, any> = yamlConfig();

      // 获取日志路径，支持环境变量覆盖
      this.logPath = process.env.LOG_PATH || config?.logs?.path || 'logs';
      this.configLoaded = true;
      Logger.log(`日志配置加载成功，路径: ${this.logPath}`);
    } catch (error: any) {
      throw new BusinessException(`加载日志配置失败: ${error.message}`);
    }
  }

  /**
   * 直接写入日志到文件
   * @param message 日志消息
   * @param level 日志级别
   */
  static writeLog(message: string, level: string = 'INFO'): void {
    // 确保配置已加载
    this.loadConfig();

    // 确保日志目录存在
    if (!fs.existsSync(this.logPath)) {
      fs.mkdirSync(this.logPath, { recursive: true });
    }

    // 日志文件名 (按日期)
    const today: string = new Date().toISOString().split('T')[0];
    const logFile: string = path.join(this.logPath, `app_${today}.log`);

    // 格式化日志消息
    const timestamp: string = new Date()
      .toISOString()
      .replace('T', ' ')
      .substring(0, 19);
    const logEntry: string = `${timestamp} - ${level} - ${message}\n`;

    // 写入文件
    try {
      fs.appendFileSync(logFile, logEntry, 'utf8');
    } catch (error: any) {
      throw new BusinessException(`写入日志失败: ${error.message}`);
    }
  }

  // 便捷静态方法
  static logInfo(message: string): void {
    Logger.log(message);
    this.writeLog(message, 'INFO');
  }

  static logError(message: string): void {
    Logger.error(message);
    this.writeLog(message, 'ERROR');
  }

  static logWarning(message: string): void {
    Logger.warn(message);
    this.writeLog(message, 'WARNING');
  }

  static logDebug(message: string): void {
    Logger.debug(message);
    this.writeLog(message, 'DEBUG');
  }

  // 获取当前日志路径
  static getLogPath(): string {
    this.loadConfig();
    return this.logPath;
  }

  // 手动重新加载配置
  static reloadConfig(): void {
    this.configLoaded = false;
    this.loadConfig();
  }
}

// SimpleLogger 类
export class SimpleLogger {
  info(message: string): void {
    LoggerUtil.logInfo(message);
  }

  error(message: string): void {
    LoggerUtil.logError(message);
  }

  warning(message: string): void {
    LoggerUtil.logWarning(message);
  }

  debug(message: string): void {
    LoggerUtil.logDebug(message);
  }
}

// 导出静态方法和实例
export const { logInfo, logError, logWarning, logDebug, writeLog } = LoggerUtil;
export const logger = new SimpleLogger();

File: gozero/app/common/utils/writeLog.go
package utils

import (
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/zeromicro/go-zero/core/logx"
)

// writeLog 直接写入日志到文件
func WriteLog(message string, level string) {
	// 使用默认路径
	logPath := filepath.Join("logs")

	// 如果日志路径不是绝对路径，则转换为绝对路径
	if !filepath.IsAbs(logPath) {
		// 获取当前工作目录
		wd, err := os.Getwd()
		if err != nil {
			logx.Error(GET_WORKING_DIR_ERROR)
		}
		// 拼接成绝对路径
		logPath = filepath.Join(wd, logPath)
	}

	// 确保日志目录存在
	if err := os.MkdirAll(logPath, 0o755); err != nil {
		logx.Error(CREATE_LOG_DIR_ERROR)
	}

	// 日志文件名 (按日期)
	today := time.Now().Format("2006-01-02")
	logFile := filepath.Join(logPath, fmt.Sprintf("app_%s.log", today))

	// 格式化日志消息
	timestamp := time.Now().Format("2006-01-02 15:04:05")
	logEntry := fmt.Sprintf("%s - %s - %s\n", timestamp, level, message)

	// 写入文件
	file, err := os.OpenFile(logFile, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o666)
	if err != nil {
		logx.Error(OPEN_LOG_FILE_ERROR)
	}
	defer file.Close()

	if _, err := file.WriteString(logEntry); err != nil {
		logx.Error(WRITE_LOG_FILE_ERROR)
	}
}

// 便捷函数
func LogInfo(message string) {
	logx.Info(message)
	WriteLog(message, "INFO")
}

func LogError(message string) {
	logx.Error(message)
	WriteLog(message, "ERROR")
}

func LogWarning(message string) {
	logx.Info(message)
	WriteLog(message, "WARNING")
}

func LogDebug(message string) {
	logx.Debug(message)
	WriteLog(message, "DEBUG")
}

// SimpleLogger 结构体
type SimpleLogger struct{}

func (l *SimpleLogger) Info(message string) {
	LogInfo(message)
}

func (l *SimpleLogger) Error(message string) {
	LogError(message)
}

func (l *SimpleLogger) Warning(message string) {
	LogWarning(message)
}

func (l *SimpleLogger) Debug(message string) {
	LogDebug(message)
}

// 创建全局logger实例
var Log = &SimpleLogger{}
